### PR TITLE
feat: custom TLS certificate for the route

### LINKED ISSUE #61: Support custom certificates for route
There should be a way to add custom certificates for the Route created by the operator. Alternatively there should be an option to disable the Route creation entirely, so we can add a custom Route.

### PR DESCRIPTION
**Pull request by mmelko**
fixes #61

### CODE REVIEW

File: pkg/controller/hawtio/hawtio_controller.go [hunk 1]
```diff
@@ -426,7 +427,19 @@ func (r *ReconcileHawtio) Reconcile(request reconcile.Request) (reconcile.Result
...
 
-	_, err = r.reconcileDeployment(hawtio, isOpenShift4, openShiftSemVer.String(), openShiftConsoleUrl, hawtio.Spec.Version, configMap, clientCertSecret)
+	tlsRouteSecret := &corev1.Secret{}
+	if secretName := hawtio.Spec.RouteTLSSecret; secretName != "" {
+		err = r.client.Get(ctx, client.ObjectKey{Namespace: request.Namespace, Name: secretName}, tlsRouteSecret)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		if tlsRouteSecret.Data["cert"] == nil || tlsRouteSecret.Data["key"] == nil {
+			return reconcile.Result{}, errosGo.New("invalid TLS route secret: cert or key field has not been found")
```
Comment (tadayosi) [inline pkg/controller/hawtio/hawtio_controller.go]: Is this typo of `errors.New`?

**Review by tadayosi — APPROVED** [review]

File: pkg/controller/hawtio/hawtio_controller.go (hunk 1 above)
Comment (mmelko) [reply pkg/controller/hawtio/hawtio_controller.go]: well it should be `errorsGo` because it's different package than errors.

Comment (tadayosi) [reply pkg/controller/hawtio/hawtio_controller.go]: Ah, ok. Then maybe `errosGo` -> `errorsGo`?

**Comment by grgrzybek** [comment]
@mmelko, can this be merged already?

File: README.md [hunk 2]
```diff
@@ -70,6 +70,29 @@ The operator covers the following cases:
   * Remove the route URL from the OAuth client authorized redirect URIs in `cluster` deployment
   * Remove the generated client certificate Secret
 
+### Custom TLS route certificate
+TLS certificate for the created route is generated by default by Openshift, however it's possible to provide 
+a own certificate stored in generic secret and point to it in the CR:
+example:
+```
+oc create secret generic route-custom-cert --from-file=cert=tls.crt --from-file=key=tls.key --from-file=caCert=caCert.crt --from-file=destinationCaCert=destinationCaCert.crt
```
Comment (astefanutti) [inline README.md]: Generally, it's recommended to use `oc create secret tls`.

File: README.md [hunk 3]
```diff
@@ -70,6 +70,29 @@ The operator covers the following cases:
   * Remove the route URL from the OAuth client authorized redirect URIs in `cluster` deployment
   * Remove the generated client certificate Secret
 
+### Custom TLS route certificate
+TLS certificate for the created route is generated by default by Openshift, however it's possible to provide 
+a own certificate stored in generic secret and point to it in the CR:
```
Comment (astefanutti) [inline README.md]: nit: "a custom certificate stored in a TLS secret, and reference it..."

File: pkg/apis/hawtio/v1alpha1/hawtio_types.go [hunk 4]
```diff
@@ -63,6 +63,9 @@ type HawtioSpec struct {
 	// Note that the operator will recreate the route if the field is emptied,
 	// so that the host is re-generated.
 	RouteHostName string `json:"routeHostName,omitempty"`
+	// Name of the generic secret with custom certificate used for route TLS
+	// Keys used in the secret: cert,key,caCert,destinationCaCert
+	RouteTLSSecret string `json:"routeTLSSecret,omitempty"`
```
Comment (astefanutti) [inline pkg/apis/hawtio/v1alpha1/hawtio_types.go]: Convention is to use `corev1.SecretKeySelector` or `corev1.LocalObjectReference`.

File: README.md [hunk 5]
```diff
@@ -70,6 +70,29 @@ The operator covers the following cases:
...
+spec:
+...
+  routeTLSSecret: route-custom-cert
+... 
+```
+The operator is looking for the following keys:
+ - `cert` - mandatory
+ - `key` - mandatory
+ - `caCert` - optional 
+ - `destinationCaCert` - optional
```
Comment (astefanutti) [inline README.md]: Why would that be used? As the operator relies on the service signing certificate service, the destination CA is injected.

Comment (mmelko) [reply README.md]: yeah so I've technically just copied what can be set when re-encript tls termination is used. It can be always blank:)

Comment (astefanutti) [reply README.md]: If it's set, then it's basically broken. It has to be injected from the service signing certificate service CA.

File: pkg/apis/hawtio/v1alpha1/hawtio_types.go [hunk 6]
```diff
@@ -63,6 +63,10 @@ type HawtioSpec struct {
 	// Note that the operator will recreate the route if the field is emptied,
 	// so that the host is re-generated.
 	RouteHostName string `json:"routeHostName,omitempty"`
+	// Name of the TLS secret with the custom certificate used for the route TLS termination
+	RouteTLSSecret corev1.LocalObjectReference `json:"routeTLSSecret,omitempty"`
+	// Name of the optional TLS secret with the CA certificate used for the route TLS termination
+	RouteTLSCaCertSecret corev1.LocalObjectReference `json:"routeTLSCaCertSecret,omitempty"`
```
Comment (astefanutti) [inline pkg/apis/hawtio/v1alpha1/hawtio_types.go]: Can this be grouped into a dedicated `route` or `tls` struct like for the other spec fields?

**Review by astefanutti — APPROVED** [review]

**Pull request merged**